### PR TITLE
Retire gp_libpq_fe

### PR DESCRIPTION
**Pull request by adam8157**
commit b0328d5631088cca5f80acc8dd85b859f062ebb0
    Author: mcdevc <a@b>
    Date:   Fri Mar 6 16:28:45 2009 -0800

        Separate our internal libpq front end from the client libpq library
        upgrade libpq to the latest to pick up bug fixes and support for more
        client authentication types (GSSAPI, KRB5, etc)
        Upgrade all files dependent on libpq to handle new version.

Above is the initial commit of gp_libpq_fe, seems no good reasons still
having it.

Key things this PR do:

1, remove the gp_libpq_fe directory.
2, build libpq source codes into two versions, for frontend and backend,
check the macro FRONTEND.
3, libpq for backend still bypasses local authentication, SSL and some
environment variables, and these are the whole differences.

Signed-off-by: Adam Lee <ali@pivotal.io>

### CODE REVIEW

**Comment by adam8157** [comment]
Some mock tests haven't been tested yet, will fix the PR CI before merging.

File: README.md [hunk 1]
```diff
@@ -296,13 +296,6 @@ throughout the codebase, but a few larger additions worth noting:
   between the DXL format used by ORCA, and the PostgreSQL internal
   representation.
 
-* __src/backend/gp_libpq_fe/__
```
Comment (hlinnaka) [inline README.md]: I think we need a new comment somewhere explaining what we do now, instead.

File: src/bin/pg_basebackup/Makefile [hunk 2]
```diff
@@ -16,7 +16,7 @@ subdir = src/bin/pg_basebackup
 top_builddir = ../../..
 include $(top_builddir)/src/Makefile.global
 
-override CPPFLAGS := -I$(libpq_srcdir) $(CPPFLAGS)
+override CPPFLAGS := -I$(libpq_srcdir) $(CPPFLAGS) $(libpq)
```
Comment (hlinnaka) [inline src/bin/pg_basebackup/Makefile]: This change shouldn't be necessary.

File: src/interfaces/libpq/exports.txt [hunk 3]
```diff
@@ -162,3 +162,17 @@ PQpingParams              159
 PQlibVersion              160
 PQsetSingleRowMode        161
 PQconninfo                162
+PQrequestFinish           163
```
Comment (hlinnaka) [inline src/interfaces/libpq/exports.txt]: Do we really need these? When linking directly into the backend, do we use these exports at all?

These numbers will actually clash with different functions that have been added to upstream libpq later. Actually, I note that PQconninfo() already has a different number than in upstream already. That could cause weird problems, if a client program uses that function.

File: src/interfaces/libpq/fe-connect.c [hunk 4]
```diff
@@ -13,20 +14,23 @@
  *-------------------------------------------------------------------------
  */
 
-#include "postgres_fe.h"
+#include "postgres.h"
```
Comment (hlinnaka) [inline src/interfaces/libpq/fe-connect.c]: Is this change from `postgres_fe.h` to `postgres.h` really needed? Can we avoid it? libpq shouldn't need to access backend functions or variables..

File: src/interfaces/libpq/fe-connect.c [hunk 5]
```diff
@@ -707,6 +739,28 @@ fillPGconn(PGconn *conn, PQconninfoOption *connOptions)
...
+		char *s = strstr(tmp, "gp_session_role");
+
+		if (s)
+		{
+			s += sizeof("gp_session_role"); /* advance to just beyond '=' */
+
+			if (s > end)
+			{
+				if (pg_strncasecmp(s, "UTILITY", sizeof("UTILITY") - 1) == 0)
+					conn->utility_mode = true;
```
Comment (hlinnaka) [inline src/interfaces/libpq/fe-connect.c]: The `utility_mode` field seems to be unused, so we can just remove this.

File: src/interfaces/libpq/fe-connect.c [hunk 6]
```diff
@@ -1479,7 +1533,12 @@ connectDBStart(PGconn *conn)
 	conn->addrlist = addrs;
 	conn->addr_cur = addrs;
 	conn->addrlist_family = hint.ai_family;
+#ifdef FOR_BACKEND
```
Comment (hlinnaka) [inline src/interfaces/libpq/fe-connect.c]: Isn't this version change a problem for anything like dblink, that tries to use the backend version of libpq to connect to a different system?

File: src/interfaces/libpq/fe-connect.c [hunk 7]
```diff
@@ -2895,12 +2968,37 @@ freePGconn(PGconn *conn)
...
+	notify = conn->notifyHead;
+	while (notify != NULL)
+	{
+		PGnotify   *prev = notify;
+
+		notify = notify->next;
+		free(prev);
+	}
+	*******/
+	pstatus = conn->pstatus;
```
Comment (hlinnaka) [inline src/interfaces/libpq/fe-connect.c]: Why do we need this now, if we didn't need it before?

File: src/interfaces/libpq/fe-exec.c [hunk 8]
```diff
@@ -1116,6 +1135,63 @@ pqRowProcessor(PGconn *conn, const char **errmsgp)
 	return 0;
 }
 
+int
```
Comment (hlinnaka) [inline src/interfaces/libpq/fe-exec.c]: I think it'd be good to have these QD-only functions in `#ifndef FRONTEND` blocks. We don't need them in libpq.so, and I don't want some client program to start using these by accident.

File: src/interfaces/libpq/fe-misc.c [hunk 9]
```diff
@@ -844,8 +901,15 @@ pqSendSome(PGconn *conn, int len)
 				case EINTR:
 					continue;
 
-				default:
-					/* pqsecure_write set the error message for us */
+				case EPIPE:
```
Comment (hlinnaka) [inline src/interfaces/libpq/fe-misc.c]: Where did these changes come from? Why was this different between the frontend and backend version?

File: src/interfaces/libpq/fe-misc.c [hunk 10]
```diff
@@ -1092,24 +1218,24 @@ pqSocketPoll(int sock, int forRead, int forWrite, time_t end_time)
 	return poll(&input_fd, 1, timeout_ms);
 #else							/* !HAVE_POLL */
 
-	fd_set		input_mask;
-	fd_set		output_mask;
-	fd_set		except_mask;
+	mpp_fd_set		input_mask;
```
Comment (hlinnaka) [inline src/interfaces/libpq/fe-misc.c]: I think we should leave out these changes, and just use the vanilla FD_* macros. AFAICS, the only difference is the that MPP_* macros can handle up to 64k fd's, while the standard glibc ones can only handle up to 1024 fds. But that's irrelevant here, this function only uses a few fds. Furthermore, we will actually use the poll() implementation on any modern platform, and this code will not be used.

File: src/interfaces/libpq/fe-protocol3.c [hunk 11]
```diff
@@ -210,6 +224,37 @@ pqParseInput3(PGconn *conn)
 							CMDSTATUS_LEN);
 					conn->asyncStatus = PGASYNC_READY;
 					break;
+				case 'o':
```
Comment (hlinnaka) [inline src/interfaces/libpq/fe-protocol3.c]: I think `#ifndef FRONTEND` would be in order here. A client program should never see these messages, and if that ever happens, we should treat that as an error. Same for all the other gpdb-specific message types, they're all only for QD->QD communication.

File: src/interfaces/libpq/libpq-fe.h [hunk 12]
```diff
@@ -196,6 +197,10 @@ typedef struct _PQconninfoOption
 								 * hide value "D"  Debug option - don't show
 								 * by default */
 	int			dispsize;		/* Field size in characters for dialog	*/
+#ifndef FRONTEND  /* modules other than backend have this macro */
```
Comment (hlinnaka) [inline src/interfaces/libpq/libpq-fe.h]: Do we still need this?

File: src/interfaces/libpq/libpq-int.h [hunk 13]
```diff
@@ -21,7 +22,7 @@
 #define LIBPQ_INT_H
 
 /* We assume libpq-fe.h has already been included. */
-#include "postgres_fe.h"
+#include "c.h"
```
Comment (hlinnaka) [inline src/interfaces/libpq/libpq-int.h]: Why this change?

File: src/interfaces/libpq/fe-connect.c [hunk 14]
```diff
@@ -251,9 +262,16 @@ static const internalPQconninfoOption PQconninfoOptions[] = {
 	 * parameters have no effect on non-SSL connections, so there is no reason
 	 * to exclude them since none of them are mandatory.
 	 */
+#ifdef FOR_BACKEND
+	/* Internal QD to QE communications don't use SSL */
+	{"sslmode", "PGSSLMODE", "disable", NULL,
```
Comment (hlinnaka) [inline src/interfaces/libpq/fe-connect.c]: Doesn't this also affect dblink, and other extensions that would like to use libpq from the backend?

**Comment by yydzero** [comment]
@pengzhout @xiong-gang any impact on dispatcher? please help to review this.

File: src/interfaces/libpq/fe-connect.c (hunk 14 above)
Comment (adam8157) [reply src/interfaces/libpq/fe-connect.c]: Yes. We disabled SSL of internal QD to QE communications for performance reason? I could create another PR to try enabling it.

File: src/interfaces/libpq/fe-connect.c (hunk 6 above)
Comment (adam8157) [reply src/interfaces/libpq/fe-connect.c]: This PR doesn't touch the fake auth part yet, it would be a problem, that's why I requested host in dblink connstr.

File: src/interfaces/libpq/libpq-int.h (hunk 13 above)
Comment (adam8157) [reply src/interfaces/libpq/libpq-int.h]: To avoid defining FRONTEND in backend.

**Comment by adam8157** [comment]
Two major issues:

Why two versions: there are still differences, local authentication, SSL and some
environment variables.

Why dblink uses backend version and has some changes: it links to backend, have to change to work with the backend version libpq.

After this, it would be perfect we figure out a way to merge this two versions.

File: src/interfaces/libpq/fe-connect.c (hunk 4 above)
Comment (adam8157) [reply src/interfaces/libpq/fe-connect.c]: #ifdef to ignore some functions in libpq.so would help.

File: src/interfaces/libpq/fe-connect.c (hunk 5 above)
Comment (danielgustafsson) [reply src/interfaces/libpq/fe-connect.c]: Agreed, this seems to be removeable.

File: src/interfaces/libpq/fe-exec.c [hunk 15]
```diff
@@ -1116,6 +1135,63 @@ pqRowProcessor(PGconn *conn, const char **errmsgp)
...
+PQsendGpQuery_shared(PGconn *conn, char *shared_query, int query_len, bool nonblock)
```
Comment (asimrp) [inline src/interfaces/libpq/fe-exec.c]: Can we dropped the `_shared` suffix?  It is causing confusion.

File: src/interfaces/libpq/fe-connect.c (hunk 7 above)
Comment (adam8157) [reply src/interfaces/libpq/fe-connect.c]: Looks it's cleaning a list, we might have a memory leak here before.

More interested is the commented out block cleaning `conn->notifyHead`, why? I just took it to avoid context missing.

File: src/interfaces/libpq/fe-exec.c (hunk 15 above)
Comment (yydzero) [reply src/interfaces/libpq/fe-exec.c]: PQsendGpQuery_shared is used by sync dispatcher, so need to dive deeper to investigate why it is introduced to support multiple threads dispatcher in first place.

Maybe worth of a separated PR.

File: src/interfaces/libpq/libpq-fe.h (hunk 12 above)
Comment (adam8157) [reply src/interfaces/libpq/libpq-fe.h]: I guess this is allowing our `psql` communicating with low version PostgreSQL?

Is it? Do we need?

File: src/interfaces/libpq/fe-connect.c (hunk 14 above)
Comment (adam8157) [reply src/interfaces/libpq/fe-connect.c]: Configuration option makes sense, maybe set it on by default.

I'd like to have another PR to give it a try, how does a GUC sound?

File: src/interfaces/libpq/fe-misc.c (hunk 9 above)
Comment (adam8157) [reply src/interfaces/libpq/fe-misc.c]: I'm not sure but backend is using this part, better to take it.

**Comment by adam8157** [comment]
Tried passing env variables like timezone to QE by GUC, however it was sent by cdbgang already.

Decide to try removing all ifdef in the next PR, possibly.
